Use Vector::uncheckedAppend() less in WebCore

cdumez opened the pull request:
#### 4a0b13037eb2952053f57e5895df92d48aea5439
<pre>
Use Vector::uncheckedAppend() less in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=262761">https://bugs.webkit.org/show_bug.cgi?id=262761</a>

Reviewed by Darin Adler.

Use Vector::uncheckedAppend() less in WebCore and replace with faster alternatives
now that uncheckedAppend() has become an alias for append().

* Source/WebCore/platform/network/HTTPHeaderMap.h:
* Source/WebCore/platform/network/cf/FormDataStreamCFNet.cpp:
(WebCore::formCreate):
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::portVectorFromList):
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::getRawCookies const):
* Source/WebCore/platform/text/LocaleICU.cpp:
(WebCore::createFallbackMonthLabels):
* Source/WebCore/platform/text/LocaleNone.cpp:
(WebCore::LocaleNone::monthLabels):
(WebCore::LocaleNone::shortMonthLabels):
* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::LocaleCocoa::monthLabels):
(WebCore::LocaleCocoa::shortMonthLabels):
(WebCore::LocaleCocoa::timeAMPMLabels):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::characterRects):
* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::computedStyleValue const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::applyStrokeStyleToContext):
* Source/WebCore/storage/Storage.cpp:
(WebCore::Storage::supportedPropertyNames const):
* Source/WebCore/style/ClassChangeInvalidation.cpp:
(WebCore::Style::collectClasses):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::updateCascadeLayerPriorities):
(WebCore::Style::RuleSetBuilder::MediaQueryCollector::pop):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertGridTrackSizeList):
(WebCore::Style::BuilderConverter::convertSVGLengthVector):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontFamily):
* Source/WebCore/style/StyleResolveForFontRaw.cpp:
(WebCore::Style::resolveForFontRaw):
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ensureInvalidationRuleSets):
* Source/WebCore/svg/SVGFEColorMatrixElement.cpp:
(WebCore::SVGFEColorMatrixElement::createFilterEffect const):
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
(WebCore::ServiceWorkerInternals::fetchResponseHeaderList):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::backgroundFetchIdentifiers):
(WebCore::BackgroundFetchEngine::matchBackgroundFetch):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::getAllResponseHeaders const):

Canonical link: <a href="https://commits.webkit.org/269027@main">https://commits.webkit.org/269027@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89bd32676411899bf6cff8bbbff2c70be073a19d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21678 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23223 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21602 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21920 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21586 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24075 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18412 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19360 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25687 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19491 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19567 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23532 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/20055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19370 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/5110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2650 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19958 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/text/LocaleICU.cpp:
```diff
@@ -188,11 +188,7 @@ std::unique_ptr<Vector<String>> LocaleICU::createLabelVector(const UDateFormat*
-    for (auto monthName : WTF::monthFullName)
-        labels->uncheckedAppend(monthName);
-    return labels;
+    return makeUnique<Vector<String>>(WTF::monthFullName, 12);
```
Why 12 instead of `std::size`? Could we do an overload that automatically does `std::size`?

cdumez replied:
Some code was using 12, some was using std::size(). I decided on 12 because this is a number of month so the size of these arrays is not going to change. I can look into the overload, or using std::size() more consistently.

darinadler replied:
If some already used 12, I am OK not pushing to use `std::size` everywhere.

darinadler commented on Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
```diff
@@ -242,10 +238,8 @@
     RetainPtr<NSDateFormatter> formatter = shortTimeFormatter();
-    m_timeAMPMLabels.uncheckedAppend([formatter AMSymbol]);
-    m_timeAMPMLabels.uncheckedAppend([formatter PMSymbol]);
+    m_timeAMPMLabels.appendList({ String([formatter AMSymbol]), String([formatter PMSymbol]) });
```
Why `append` when this is known to be empty? Could we assign instead?

cdumez replied:
Sure, will fix.

darinadler commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -544,11 +544,10 @@ static Vector<LayoutRect> characterRects(const InlineIterator::TextBox& run, uns
-        for (auto index = clampedStart; index < clampedEnd; ++index)
-            rects.uncheckedAppend(svgTextBox->localSelectionRect(index, index + 1));
-        return rects;
+        return Vector<LayoutRect>(clampedEnd - clampedStart, [&, clampedStart = clampedStart](size_t i) {
```
Why do we need to write `clampedStart = clampedStart`? Doesn’t `&` already capture `clampedStart`?

cdumez replied:
It is because `clampedStart` is initialized like so (structured bindings):
```
auto [clampedStart, clampedEnd] = run.selectableRange().clamp(rangeStart, rangeEnd);
```

I was getting a build error until I captured it explicitly by value.

cdumez replied:
To clarify, it is building locally for me with just [&] and that's what I originally had. However, quite a few of the bots didn't like it (both iOS and macOS):
```
/Volumes/Data/worker/macOS-AppleSilicon-Ventura-Debug-Build-EWS/build/Source/WebCore/rendering/RenderText.cpp:548:28: error: reference to local binding 'clampedStart' declared in enclosing function 'WebCore::characterRects'
```
from https://ews-build.webkit.org/#/builders/30/builds/19242

darinadler replied:
Then I think the correct way to say this is something like "`clampedStart = clampedStart` is a workaround for what seems to be a compiler bug in older versions of `clang`".

cdumez replied:
I did some research and found https://en.cppreference.com/w/cpp/language/structured_binding:
```
Structured bindings cannot be captured by lambda expressions (until C++20)
```

So it seems this is only supported since C++20 and it seems some of our toolchains don't support this particular C++20 feature yet.

darinadler commented on Source/WebCore/svg/SVGFEColorMatrixElement.cpp:
```diff
@@ -127,9 +127,9 @@ RefPtr<FilterEffect> SVGFEColorMatrixElement::createFilterEffect(const FilterEff
-            for (size_t i = 0; i < matrixValueCount; i++)
-                filterValues.uncheckedAppend((i % 6) ? 0 : 1);
+            filterValues = Vector<float>(matrixValueCount, [](size_t i) {
+                return (i % 6) ? 0. : 1.;
```
I know `0.` and `1.` work, but `0.0` and `1.0` are more traditional, I think. Or we could use `-> double`.

cdumez replied:
Oops. I had missed this review comment. Will fix in next iteration.

cdumez committed:
Use Vector::uncheckedAppend() less in WebCore
https://bugs.webkit.org/show_bug.cgi?id=262761

Reviewed by Darin Adler.

Use Vector::uncheckedAppend() less in WebCore and replace with faster alternatives
now that uncheckedAppend() has become an alias for append().

* Source/WebCore/platform/network/HTTPHeaderMap.h:
* Source/WebCore/platform/network/cf/FormDataStreamCFNet.cpp:
(WebCore::formCreate):
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::portVectorFromList):
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::getRawCookies const):
* Source/WebCore/platform/text/LocaleICU.cpp:
(WebCore::createFallbackMonthLabels):
* Source/WebCore/platform/text/LocaleNone.cpp:
(WebCore::LocaleNone::monthLabels):
(WebCore::LocaleNone::shortMonthLabels):
* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::LocaleCocoa::monthLabels):
(WebCore::LocaleCocoa::shortMonthLabels):
(WebCore::LocaleCocoa::timeAMPMLabels):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::characterRects):
* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::computedStyleValue const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::applyStrokeStyleToContext):
* Source/WebCore/storage/Storage.cpp:
(WebCore::Storage::supportedPropertyNames const):
* Source/WebCore/style/ClassChangeInvalidation.cpp:
(WebCore::Style::collectClasses):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::updateCascadeLayerPriorities):
(WebCore::Style::RuleSetBuilder::MediaQueryCollector::pop):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertGridTrackSizeList):
(WebCore::Style::BuilderConverter::convertSVGLengthVector):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontFamily):
* Source/WebCore/style/StyleResolveForFontRaw.cpp:
(WebCore::Style::resolveForFontRaw):
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ensureInvalidationRuleSets):
* Source/WebCore/svg/SVGFEColorMatrixElement.cpp:
(WebCore::SVGFEColorMatrixElement::createFilterEffect const):
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
(WebCore::ServiceWorkerInternals::fetchResponseHeaderList):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::backgroundFetchIdentifiers):
(WebCore::BackgroundFetchEngine::matchBackgroundFetch):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::getAllResponseHeaders const):

Canonical link: https://commits.webkit.org/269027@main
